app/vmctl/opentsdb/opentsdb.go

```diff
@@ -0,0 +1,279 @@
+	Expressions []int       `json:"expressions"`
+	Outputs     []outputObj `json:"outputs"`
+}
+
+type timeObj struct {
+	Start       int64  `json:"start"`
+	End         int64  `json:"end"`
+	Aggregator  string `json:"aggregator"`
+	Downsampler dSObj  `json:"downsampler"`
+}
+
+type dSObj struct {
+	Interval   string  `json:"interval"`
+	Aggregator string  `json:"aggregator"`
+	FillPolicy fillObj `json:"fillPolicy"`
+}
+
+type fillObj struct {
+	// we'll always hard-code to NaN here, so we don't need value
+	Policy string `json:"policy"`
+}
+
+type filterObj struct {
+	Tags []tagObj `json:"tags"`
+	ID   string   `json:"id"`
+}
+
+type tagObj struct {
+	Type    string `json:"type"`
+	Tagk    string `json:"tagk"`
+	Filter  string `json:"filter"`
+	GroupBy bool   `json:"groupBy"`
+}
+
+type metricObj struct {
+	ID         string  `json:"id"`
+	Metric     string  `json:"metric"`
+	Filter     string  `json:"filter"`
+	FillPolicy fillObj `json:"fillPolicy"`
+}
+
+type outputObj struct {
+	ID    string `json:"id"`
+	Alias string `json:"alias"`
+}
+
+/* End expression object structs */
+
+// FindMetrics discovers all metrics that OpenTSDB knows about (given a filter)
+// e.g. /api/suggest?type=metrics&q=system&max=100000
+func (c Client) FindMetrics(filter string) ([]string, error) {
+	q := fmt.Sprintf("%s/api/suggest?type=metrics&q=%s&max=%d", c.Addr, filter, c.Limit)
+	//log.Println(q)
+	resp, err := http.Get(q)
+	if err != nil {
+		return nil, fmt.Errorf("Could not properly make request to %s: %s", c.Addr, err)
+	}
+	defer func() { _ = resp.Body.Close() }()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Could not retrieve metric data from %s: %s", c.Addr, err)
+	}
+	var metriclist []string
+	err = json.Unmarshal(body, &metriclist)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid metric data from %s: %s", c.Addr, err)
+	}
+	return metriclist, nil
+}
+
+// FindSeries discovers all series associated with a metric
+// e.g. /api/search/lookup?m=system.load5&limit=1000000
+func (c Client) FindSeries(metric string) ([]Meta, error) {
+	q := fmt.Sprintf("%s/api/search/lookup?m=%s&limit=%d", c.Addr, metric, c.Limit)
+	resp, err := http.Get(q)
+	if err != nil {
+		return nil, fmt.Errorf("Could not properly make request to %s: %s", c.Addr, err)
+	}
+	defer func() { _ = resp.Body.Close() }()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Could not retrieve series data from %s: %s", c.Addr, err)
+	}
+	var results MetaResults
+	err = json.Unmarshal(body, &results)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid series data from %s: %s", c.Addr, err)
+	}
+	return results.Results, nil
+}
+
+// GetData actually retrieves data for a series at a specified time range
+func (c Client) GetData(series Meta, rt Retention, start int64, end int64) (Metric, error) {
+	expr := Expression{}
+	expr.Outputs = append(expr.Outputs, outputObj{ID: "a", Alias: "query"})
```

Comment:
there is no need to do `append` since `outputObj{ID: "a", Alias: "query"}` is the only element in the slice.
You also can define static objects as constants instead of re-creating them on each call:
```
var (
	exprOutput     = outputObj{ID: "a", Alias: "query"}
	exprFillPolicy = fillObj{Policy: "nan"}
)

// GetData actually retrieves data for a series at a specified time range
func (c Client) GetData(series Meta, rt Retention, start int64, end int64) (Metric, error) {
	expr := Expression{
		Outputs: []outputObj{exprOutput},
		Metrics: []metricObj{
			{
				ID: "a", Metric: series.Metric,
				Filter: "f1", FillPolicy: exprFillPolicy,
			},
		},
		Time: timeObj{
			Start: start, End: end, Aggregator: rt.FirstOrder,
			Downsampler: dSObj{
				Interval:   rt.AggTime,
				Aggregator: rt.SecondOrder,
				FillPolicy: exprFillPolicy},
		},
	}
```